keps/sig-node/2823-pod-addmission-plugin/README.md

```diff
@@ -0,0 +1,445 @@
+and [loading](https://github.com/kubernetes/kubernetes/blob/dd5272b76f07bea60628af0bb793f3cca385bf5e/pkg/kubelet/dockershim/docker_service.go#L248)
+plugin [descriptors](https://github.com/kubernetes/kubernetes/blob/f4db8212be53c69a27d893d6a4111422fbce8008/pkg/kubelet/dockershim/network/plugins.go#L52)
+upon startup.
+
+To support pluggable validation for pod admission on the worker node, we propose
+to have kubelet similarly discover node-local Pod admission plugins listed in a
+new PodAdmissionPluginDir flag.
+
+Other option is to enable this functionality through feature flag
+“enablePodAdmissionPlugin” and have the directory path defined inside the
+kubelet itself.
+
+### User Stories (Optional)
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+These out-of-tree plugins should be provisioned and managed along with kubelet.
+
+## Design Details
+
+### Design
+
+#### Configuration file
+
+Node-local Pod admission plugins will be listed in a configuration file. The
+Plugins field indicates the list of plugins to invoke before admitting the Pod.
+
+```
+{
+  "name": "admission-plugin",
+  "version": "0.1",
+  "plugins": [
+    {
+      "name": "sysctlcheck",
+      "type": "shell"
+    },
+    {
+      "name": "fargatecheck",
+      "type": "shell"
+    }
+  ]
+}
+```
+
+A node-local Pod admission plugin has the following structure:
+
+```
+
+// PluginType indicates type of the admission plugin
+type PluginType string
+
+const (
+    PluginTypeShell  PluginType = "shell"  // binary to execute.
+    PluginTypeGRPC   PluginType = "grpc"   // Local port on the host.
+    PluginTypeSocket PluginType = "socket" // fd to connect to.
+)
+
+// AdmissionPluginManager is the podAdmitHandler shim for external plugins.
+type AdmissionPluginManager struct {
+    confDir      string
+    binDir       string
+    pluginConfig *PluginConfig
+}
+
+// PluginConfig represents the plugin configuration file
+type PluginConfig struct {
+    Name         string              `json:"name,omitempty"`
+    Version      string              `json:"version,omitempty"`
+    Plugins      []*AdmissionPlugin  `json:"plugins"`
+}
+
+// AdmissionPlugin represents individual plugins specified in the configuration.
+type AdmissionPlugin struct {
+    Name     string     `json:"name"`
+    Type     PluginType `json:"type"`
+}
+
+```
+
+#### Feature gate
+
+This functionality adds a new feature gate named “PodAdmissionPlugin” which
+decides whether to invoke admission plugin or not.
+
+#### Kubelet to pod admission plugin communication
+
+Kubelet will encode the pod spec and invoke each admission plugin. After
+decoding the pod spec, plugins can perform additional validations and return the
+encoded form of the struct mentioned below to kubelet to decide whether to admit
+the pod or not. For shell plugin type, request will be sent over stdin and
+response will be received over stdout and stderr.
+
+```
+AdmitResult { 
```

Comment:
I think we should capture how the rejection manifests itself from user perspective.  e.g. a Pod has a status Reason ( `CustomAdmissionFailure`) along with description with the `Message`. Lets explain this with concrete examples. It is implicit but it might be work mentioning that this allows each plugin to have mechanism to have a more plugin specific pod rejection error message.

Reply:
Definitely, we will add diagrams and examples in user stories.